In [2]:
import sys
print(sys.executable)

/opt/anaconda3/envs/cell-aging/bin/python3.11


In [3]:
from pathlib import Path
import pandas as pd
import numpy as np

PROJECT_ROOT = Path.cwd().parent
print(PROJECT_ROOT)

/Users/zhangyuxiang/Desktop/cell-aging-analysis


In [4]:
data_path = PROJECT_ROOT / "data" / "processed" / "df_clean_wide.csv"

print(data_path)
print("Exists:", data_path.exists())

/Users/zhangyuxiang/Desktop/cell-aging-analysis/data/processed/df_clean_wide.csv
Exists: True


In [5]:
df = pd.read_csv(data_path)
df.shape

(113197, 26)

In [6]:
df.head()

,ImageNumber,Count_CellsIncludingEdges,Count_Cytoplasm,Count_Nuclei,Count_NucleiIncludingEdges,Metadata_Plate,Metadata_Well,Metadata_JCP2022,Cells_AreaShape_BoundingBoxArea,Cells_AreaShape_Compactness,...,Nuclei_AreaShape_BoundingBoxArea,Nuclei_AreaShape_Compactness,Nuclei_AreaShape_Eccentricity,Nuclei_AreaShape_EulerNumber,Nuclei_AreaShape_FormFactor,Nuclei_AreaShape_MeanRadius,Nuclei_AreaShape_Perimeter,Nuclei_AreaShape_Solidity,NC_Ratio,NC_Ratio_Radius
0,14014510,247,213,213,247,CP1-SC1-21,A22,JCP2022_095995,738.5300,2.549574,...,285.6100,6.366034,0.253077,1,0.157084,0.948338,86.034933,0.498861,0.386728,0.369107
1,14014510,247,213,213,247,CP1-SC1-21,A22,JCP2022_095995,1064.7000,1.735442,...,635.8625,2.594600,0.424371,1,0.385416,3.257558,110.538948,0.810046,0.597222,0.753967
2,14014510,247,213,213,247,CP1-SC1-21,A22,JCP2022_095995,615.5825,2.819940,...,312.6500,5.335602,0.871403,1,0.187420,1.203673,91.415747,0.612033,0.507893,0.544603
3,14014510,247,213,213,247,CP1-SC1-21,A22,JCP2022_095995,202.8000,3.821175,...,202.8000,3.821175,0.786616,1,0.261700,1.136370,68.011784,0.668622,1.000000,1.000000
4,14014510,247,213,213,247,CP1-SC1-21,A22,JCP2022_095995,1324.5375,2.743161,...,1254.8250,3.526792,0.521716,1,0.283544,3.540678,170.746769,0.739316,0.947368,0.919785


In [7]:
df.columns.tolist()

['ImageNumber',
 'Count_CellsIncludingEdges',
 'Count_Cytoplasm',
 'Count_Nuclei',
 'Count_NucleiIncludingEdges',
 'Metadata_Plate',
 'Metadata_Well',
 'Metadata_JCP2022',
 'Cells_AreaShape_BoundingBoxArea',
 'Cells_AreaShape_Compactness',
 'Cells_AreaShape_Eccentricity',
 'Cells_AreaShape_EulerNumber',
 'Cells_AreaShape_FormFactor',
 'Cells_AreaShape_MeanRadius',
 'Cells_AreaShape_Perimeter',
 'Cells_AreaShape_Solidity',
 'Nuclei_AreaShape_BoundingBoxArea',
 'Nuclei_AreaShape_Compactness',
 'Nuclei_AreaShape_Eccentricity',
 'Nuclei_AreaShape_EulerNumber',
 'Nuclei_AreaShape_FormFactor',
 'Nuclei_AreaShape_MeanRadius',
 'Nuclei_AreaShape_Perimeter',
 'Nuclei_AreaShape_Solidity',
 'NC_Ratio',
 'NC_Ratio_Radius']

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 113197 entries, 0 to 113196
Data columns (total 26 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   ImageNumber                       113197 non-null  int64  
 1   Count_CellsIncludingEdges         113197 non-null  int64  
 2   Count_Cytoplasm                   113197 non-null  int64  
 3   Count_Nuclei                      113197 non-null  int64  
 4   Count_NucleiIncludingEdges        113197 non-null  int64  
 5   Metadata_Plate                    113197 non-null  str    
 6   Metadata_Well                     113197 non-null  str    
 7   Metadata_JCP2022                  113197 non-null  str    
 8   Cells_AreaShape_BoundingBoxArea   113197 non-null  float64
 9   Cells_AreaShape_Compactness       113197 non-null  float64
 10  Cells_AreaShape_Eccentricity      113197 non-null  float64
 11  Cells_AreaShape_EulerNumber       113197 non-null  int64  
 12 

In [9]:
metadata_cols = [c for c in df.columns if c.startswith("Metadata")]
feature_cols = [c for c in df.columns if not c.startswith("Metadata")]

print("Number of metadata columns:", len(metadata_cols))
print("Number of feature columns:", len(feature_cols))

print("\nMetadata columns:")
print(metadata_cols)

print("\nFeature columns:")
print(feature_cols)

Number of metadata columns: 3
Number of feature columns: 23

Metadata columns:
['Metadata_Plate', 'Metadata_Well', 'Metadata_JCP2022']

Feature columns:
['ImageNumber', 'Count_CellsIncludingEdges', 'Count_Cytoplasm', 'Count_Nuclei', 'Count_NucleiIncludingEdges', 'Cells_AreaShape_BoundingBoxArea', 'Cells_AreaShape_Compactness', 'Cells_AreaShape_Eccentricity', 'Cells_AreaShape_EulerNumber', 'Cells_AreaShape_FormFactor', 'Cells_AreaShape_MeanRadius', 'Cells_AreaShape_Perimeter', 'Cells_AreaShape_Solidity', 'Nuclei_AreaShape_BoundingBoxArea', 'Nuclei_AreaShape_Compactness', 'Nuclei_AreaShape_Eccentricity', 'Nuclei_AreaShape_EulerNumber', 'Nuclei_AreaShape_FormFactor', 'Nuclei_AreaShape_MeanRadius', 'Nuclei_AreaShape_Perimeter', 'Nuclei_AreaShape_Solidity', 'NC_Ratio', 'NC_Ratio_Radius']


In [11]:
"Metadata_JCP2022" in df.columns
df["Metadata_JCP2022"].value_counts().head(20)

Metadata_JCP2022
JCP2022_033924    93109
JCP2022_095995     7150
JCP2022_001418     6143
JCP2022_051785     4739
JCP2022_094347     2056
Name: count, dtype: int64

In [12]:
missing_rate = df.isna().mean().sort_values(ascending=False)
missing_rate

ImageNumber                         0.0
Count_CellsIncludingEdges           0.0
NC_Ratio                            0.0
Nuclei_AreaShape_Solidity           0.0
Nuclei_AreaShape_Perimeter          0.0
Nuclei_AreaShape_MeanRadius         0.0
Nuclei_AreaShape_FormFactor         0.0
Nuclei_AreaShape_EulerNumber        0.0
Nuclei_AreaShape_Eccentricity       0.0
Nuclei_AreaShape_Compactness        0.0
Nuclei_AreaShape_BoundingBoxArea    0.0
Cells_AreaShape_Solidity            0.0
Cells_AreaShape_Perimeter           0.0
Cells_AreaShape_MeanRadius          0.0
Cells_AreaShape_FormFactor          0.0
Cells_AreaShape_EulerNumber         0.0
Cells_AreaShape_Eccentricity        0.0
Cells_AreaShape_Compactness         0.0
Cells_AreaShape_BoundingBoxArea     0.0
Metadata_JCP2022                    0.0
Metadata_Well                       0.0
Metadata_Plate                      0.0
Count_NucleiIncludingEdges          0.0
Count_Nuclei                        0.0
Count_Cytoplasm                     0.0


## Initial observations

The cleaned dataset contains 113,197 rows and 26 columns. There are 3 metadata columns: `Metadata_Plate`, `Metadata_Well`, and `Metadata_JCP2022`. The remaining columns include object counts and selected CellProfiler AreaShape features for cells and nuclei, as well as derived N/C ratio variables.

There are no missing values in the current cleaned dataset. The compound identifier column is `Metadata_JCP2022`, with 5 unique compound IDs in this file.

In [13]:
df["Metadata_JCP2022"].unique()
compound_counts = (
    df["Metadata_JCP2022"]
    .value_counts()
    .rename_axis("Metadata_JCP2022")
    .reset_index(name="n_cells")
)

compound_counts

,Metadata_JCP2022,n_cells
0,JCP2022_033924,93109
1,JCP2022_095995,7150
2,JCP2022_001418,6143
3,JCP2022_051785,4739
4,JCP2022_094347,2056


In [14]:
numeric_cols = df.select_dtypes(include=["number"]).columns.tolist()

summary_table = df[numeric_cols].describe().T
summary_table

,count,mean,std,min,25%,50%,75%,max
ImageNumber,113197.0,1.113698e+07,3.316233e+06,12937.000000,1.201253e+07,1.202050e+07,1.202463e+07,1.401452e+07
Count_CellsIncludingEdges,113197.0,3.166048e+02,6.871720e+01,104.000000,2.760000e+02,3.190000e+02,3.570000e+02,5.150000e+02
Count_Cytoplasm,113197.0,2.751987e+02,6.505386e+01,78.000000,2.360000e+02,2.770000e+02,3.130000e+02,4.710000e+02
Count_Nuclei,113197.0,2.752012e+02,6.505414e+01,78.000000,2.360000e+02,2.770000e+02,3.130000e+02,4.710000e+02
Count_NucleiIncludingEdges,113197.0,3.166048e+02,6.871720e+01,104.000000,2.760000e+02,3.190000e+02,3.570000e+02,5.150000e+02
Cells_AreaShape_BoundingBoxArea,113197.0,2.775736e+03,1.843471e+03,2.535000,1.647750e+03,2.499932e+03,3.549000e+03,4.379804e+04
Cells_AreaShape_Compactness,113197.0,2.734732e+00,1.091309e+00,0.327399,2.086901e+00,2.470849e+00,3.031921e+00,2.088859e+01
Cells_AreaShape_Eccentricity,113197.0,7.774714e-01,1.397727e-01,0.016798,7.033215e-01,8.094690e-01,8.829008e-01,9.960543e-01
Cells_AreaShape_EulerNumber,113197.0,9.321272e-01,3.317420e-01,-17.000000,1.000000e+00,1.000000e+00,1.000000e+00,4.000000e+00
Cells_AreaShape_FormFactor,113197.0,4.038311e-01,1.116883e-01,0.047873,3.298239e-01,4.047192e-01,4.791794e-01,3.054380e+00


In [15]:
selected_features = [
    "Cells_AreaShape_BoundingBoxArea",
    "Cells_AreaShape_MeanRadius",
    "Cells_AreaShape_Perimeter",
    "Cells_AreaShape_Solidity",
    "Nuclei_AreaShape_BoundingBoxArea",
    "Nuclei_AreaShape_MeanRadius",
    "Nuclei_AreaShape_Perimeter",
    "Nuclei_AreaShape_Solidity",
    "NC_Ratio",
    "NC_Ratio_Radius",
]

df[selected_features].describe().T

,count,mean,std,min,25%,50%,75%,max
Cells_AreaShape_BoundingBoxArea,113197.0,2775.736299,1843.470642,2.535000,1647.750000,2499.932500,3549.000000,43798.040000
Cells_AreaShape_MeanRadius,113197.0,4.801459,1.511393,0.650000,4.087230,4.918226,5.687271,21.504030
Cells_AreaShape_Perimeter,113197.0,206.910887,77.577992,3.384619,159.004538,201.035456,247.666531,1219.774992
Cells_AreaShape_Solidity,113197.0,0.787973,0.086148,0.016529,0.743972,0.805435,0.849945,1.000000
Nuclei_AreaShape_BoundingBoxArea,113197.0,521.816445,263.058894,93.372500,385.320000,477.847500,591.500000,5635.305000
Nuclei_AreaShape_MeanRadius,113197.0,3.053898,0.750243,0.713581,2.829839,3.195358,3.520391,6.342415
Nuclei_AreaShape_Perimeter,113197.0,81.763322,31.979213,31.607821,66.734881,74.469553,84.331075,554.344480
Nuclei_AreaShape_Solidity,113197.0,0.899211,0.114041,0.284487,0.900952,0.950382,0.962806,0.985102
NC_Ratio,113197.0,0.308302,1.614005,0.005954,0.139808,0.193266,0.289435,433.333333
NC_Ratio_Radius,113197.0,0.691292,0.299064,0.068927,0.567722,0.648633,0.750313,6.795044


In [16]:
df[["NC_Ratio", "NC_Ratio_Radius"]].describe()

,NC_Ratio,NC_Ratio_Radius
count,113197.000000,113197.000000
mean,0.308302,0.691292
std,1.614005,0.299064
min,0.005954,0.068927
25%,0.139808,0.567722
50%,0.193266,0.648633
75%,0.289435,0.750313
max,433.333333,6.795044


In [17]:
group_summary = df.groupby("Metadata_JCP2022")[selected_features].mean()
group_summary

,Cells_AreaShape_BoundingBoxArea,Cells_AreaShape_MeanRadius,Cells_AreaShape_Perimeter,Cells_AreaShape_Solidity,Nuclei_AreaShape_BoundingBoxArea,Nuclei_AreaShape_MeanRadius,Nuclei_AreaShape_Perimeter,Nuclei_AreaShape_Solidity,NC_Ratio,NC_Ratio_Radius
Metadata_JCP2022,,,,,,,,,,
JCP2022_001418,3957.213406,5.405022,254.771783,0.771013,597.906621,3.017492,91.530011,0.862472,0.443248,0.763365
JCP2022_033924,2632.637468,4.743038,201.286537,0.790122,513.810042,3.052936,80.667898,0.902936,0.293522,0.679482
JCP2022_051785,2818.760415,4.841249,210.136928,0.789090,566.278471,3.047338,88.734276,0.875777,0.350374,0.728522
JCP2022_094347,5660.930981,5.651454,308.646016,0.742669,542.826808,3.008793,86.148511,0.878788,0.573243,0.941590
JCP2022_095995,2765.963377,4.772875,207.639904,0.786846,525.193145,3.115020,81.755738,0.903679,0.280767,0.686512
